crates/network/src/rpc_router.rs

```diff
@@ -0,0 +1,412 @@
+pub struct RpcRouter<T>
+where
+    T: RpcRequest,
+{
+    networking: Networking,
+    response_tracker: ResponseTracker<T::Response>,
+}
+
+#[derive(thiserror::Error, Debug)]
+#[error(transparent)]
+pub enum RpcError {
+    #[error("correlation id {0} is already in-flight")]
+    CorrelationIdExists(String),
+    SendError(#[from] NetworkSendError),
+    Shutdown(#[from] ShutdownError),
+}
+
+impl<T> RpcRouter<T>
+where
+    T: RpcRequest + WireEncode + Send + Sync + 'static,
+    T::Response: WireDecode + Send + Sync + 'static,
+    <T::Response as RpcMessage>::CorrelationId: Send + Sync + From<T::CorrelationId>,
+{
+    pub fn new(networking: Networking, router_builder: &mut MessageRouterBuilder) -> Self {
+        let response_tracker = ResponseTracker::<T::Response>::default();
+        router_builder.add_message_handler(response_tracker.clone());
+        Self {
+            networking,
+            response_tracker,
+        }
+    }
+
+    pub async fn call(&self, to: NodeId, msg: &T) -> Result<MessageEnvelope<T::Response>, RpcError>
+    where
+        <T::Response as RpcMessage>::CorrelationId: Default,
+    {
+        let token = self
+            .response_tracker
+            .new_token(msg.correlation_id().into())
+            .ok_or_else(|| RpcError::CorrelationIdExists(format!("{:?}", msg.correlation_id())))?;
+        self.networking.send(to, &msg).await?;
+        token
+            .recv()
+            .await
+            .map_err(|_| RpcError::Shutdown(ShutdownError))
+    }
+
+    pub fn num_in_flight(&self) -> usize {
+        self.response_tracker.num_in_flight()
+    }
+}
+
+/// A tracker for responses but can be used to track responses for requests that were dispatched
+/// via other mechanisms (e.g. ingress flow)
+pub struct ResponseTracker<T>
+where
+    T: RpcMessage,
+{
+    inner: Arc<Inner<T>>,
+}
+
+impl<T> Clone for ResponseTracker<T>
+where
+    T: RpcMessage,
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+struct Inner<T>
+where
+    T: RpcMessage,
+{
+    in_flight: DashMap<T::CorrelationId, RpcTokenSender<T>>,
+}
+
+impl<T> Default for ResponseTracker<T>
+where
+    T: RpcMessage,
+{
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(Inner {
+                in_flight: Default::default(),
+            }),
+        }
+    }
+}
+
+impl<T> ResponseTracker<T>
+where
+    T: RpcMessage,
+{
+    pub fn num_in_flight(&self) -> usize {
+        self.inner.in_flight.len()
+    }
+
+    /// Returns None if an in-flight request holds the same correlation_id.
+    pub fn new_token(&self, correlation_id: T::CorrelationId) -> Option<RpcToken<T>> {
+        let (sender, receiver) = oneshot::channel();
+        let existing = self
+            .inner
+            .in_flight
+            .insert(correlation_id.clone(), RpcTokenSender { sender });
+
+        if existing.is_some() {
+            return None;
```

Comment:
I'll add this but it won't guarantee correctness anyway. I'll add a warn message though.

Reply:
Ah true, because we are not holding an exclusive lock over `Inner`. Damn. Then I guess adding the token sender back is actually even more dangerous because as you've said we might have missed the signal and will potentially be forever stuck waiting on it.

Reply:
I think we could use `in_flight.entry()` to make it work correctly because `entry()` keeps the lock as long as it exists.